frontend/test/pytest/test_autograph.py

```diff
@@ -432,5 +506,327 @@ def circuit():
+        assert np.allclose(result, -jnp.sqrt(2) / 2)
+
+    def test_for_in_array_unpack(self):
+        """Test for loop over a 2D JAX array unpacking the inner dimension."""
+
+        @qjit(autograph=True)
+        @qml.qnode(qml.device("lightning.qubit", wires=1))
+        def f(params):
+            for x1, x2 in params:
+                qml.RY(x1, wires=0)
+                qml.RY(x2, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        result = f(jnp.array([[0.0, 1 / 4 * jnp.pi], [2 / 4 * jnp.pi, jnp.pi]]))
+        assert np.allclose(result, jnp.sqrt(2) / 2)
+
+    def test_for_in_numeric_list(self):
+        """Test for loop over a Python list that is convertible to an array."""
+
+        @qjit(autograph=True)
+        @qml.qnode(qml.device("lightning.qubit", wires=1))
+        def f():
+            params = [0.0, 1 / 4 * jnp.pi, 2 / 4 * jnp.pi]
+            for x in params:
+                qml.RY(x, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        result = f()
+        assert np.allclose(result, -jnp.sqrt(2) / 2)
+
+    def test_for_in_numeric_list_of_list(self):
+        """Test for loop over a nested Python list that is convertible to an array."""
+
+        @qjit(autograph=True)
+        @qml.qnode(qml.device("lightning.qubit", wires=1))
+        def f():
+            params = [[0.0, 1 / 4 * jnp.pi], [2 / 4 * jnp.pi, jnp.pi]]
+            for xx in params:
+                for x in xx:
+                    qml.RY(x, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        result = f()
+        assert np.allclose(result, jnp.sqrt(2) / 2)
+
+    def test_for_in_object_list(self):
+        """Test for loop over a Python list that is *not* convertible to an array.
+        The behaviour should fall back to standard Python."""
+
+        @qjit(autograph=True)
+        @qml.qnode(qml.device("lightning.qubit", wires=1))
+        def f():
+            params = ["0", "1", "2"]
+            for x in params:
+                qml.RY(int(x) / 4 * jnp.pi, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        result = f()
+        assert np.allclose(result, -jnp.sqrt(2) / 2)
+
+    def test_for_in_static_range(self):
+        """Test for loop over a Python range with static bounds."""
+
+        @qjit(autograph=True)
+        @qml.qnode(qml.device("lightning.qubit", wires=3))
+        def f():
+            for i in range(3):
+                qml.Hadamard(i)
+            return qml.probs()
+
+        result = f()
+        assert np.allclose(result, [1 / 8] * 8)
+
+    def test_for_in_static_range_indexing_array(self):
+        """Test for loop over a Python range with static bounds that is used to index an array."""
+
+        @qml.qnode(qml.device("lightning.qubit", wires=1))
+        def f():
+            params = jnp.array([0.0, 1 / 4 * jnp.pi, 2 / 4 * jnp.pi])
+            for i in range(3):
+                qml.RY(params[i], wires=0)
+            return qml.expval(qml.PauliZ(0))
```

Comment:
Is this one falling back to python?

Reply:
No with this one the conversion succeeds :)